scipy/optimize/_qap.py

```diff
@@ -0,0 +1,780 @@
+    if isinstance(P0, str) and P0 not in {'barycenter', 'randomized'}:
+        msg = "Invalid 'P0' parameter string"
+    elif maxiter <= 0:
+        msg = "'maxiter' must be a positive integer"
+    elif tol <= 0:
+        msg = "'tol' must be a positive float"
+    if msg is not None:
+        raise ValueError(msg)
+
+    rng = check_random_state(rng)
+    n = A.shape[0]  # number of vertices in graphs
+    n_seeds = partial_match.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+
+    nonseed_B = np.setdiff1d(range(n), partial_match[:, 1])
+    if shuffle_input:
+        nonseed_B = rng.permutation(nonseed_B)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    nonseed_A = np.setdiff1d(range(n), partial_match[:, 0])
+    perm_A = np.concatenate([partial_match[:, 0], nonseed_A])
+    perm_B = np.concatenate([partial_match[:, 1], nonseed_B])
+
+    # definitions according to Seeded Graph Matching [2].
+    A11, A12, A21, A22 = _split_matrix(A[perm_A][:, perm_A], n_seeds)
+    B11, B12, B21, B22 = _split_matrix(B[perm_B][:, perm_B], n_seeds)
+
+    # [1] Algorithm 1 Line 1 - choose initialization
+    if isinstance(P0, str):
+        # initialize J, a doubly stochastic barycenter
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+        if P0 == 'barycenter':
+            P = J
+        elif P0 == 'randomized':
+            # generate a nxn matrix where each entry is a random number [0, 1]
+            # would use rand, but Generators don't have it
+            # would use random, but old mtrand.RandomStates don't have it
+            K = rng.uniform(size=(n_unseed, n_unseed))
+            # Sinkhorn balancing
+            K = _doubly_stochastic(K)
+            P = J * 0.5 + K * 0.5
+    else:
+        P0 = np.atleast_2d(P0)
+        _check_init_input(P0, n_unseed)
+        P = P0
+
+    const_sum = A21 @ B21.T + A12.T @ B12
+
+    # [1] Algorithm 1 Line 2 - loop while stopping criteria not met
+    for n_iter in range(1, maxiter+1):
+        # [1] Algorithm 1 Line 3 - compute the gradient of f(P) = -tr(APB^tP^t)
+        grad_fp = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+        # [1] Algorithm 1 Line 4 - get direction Q by solving Eq. 8
+        _, cols = linear_sum_assignment(grad_fp, maximize=maximize)
+        Q = np.eye(n_unseed)[cols]
+
+        # [1] Algorithm 1 Line 5 - compute the step size
+        # Noting that e.g. trace(Ax) = trace(A)*x, expand and re-collect
+        # terms as ax**2 + bx + c. c does not affect location of minimum
+        # and can be ignored. Also, note that trace(A@B) = (A.T*B).sum();
+        # apply where possible for efficiency.
+        R = P - Q
+        b21 = ((R.T @ A21) * B21).sum()
+        b12 = ((R.T @ A12.T) * B12.T).sum()
+        AR22 = A22.T @ R
+        BR22 = B22 @ R.T
+        b22a = (AR22.T * (B22 @ Q.T)).sum()
+        b22b = ((A22.T @ Q).T * BR22).sum()
```

Comment:
Since `Q` is a permutation matrix we can shave off a couple of matmul operations:
```
        b22a = (AR22 * B22.T[cols]).sum()
        b22b = (A22 * BR22[cols]).sum()
```
Its about half the cost at larger matrix sizes.